modules/lsm/lsm_local.c

```diff
@@ -0,0 +1,256 @@
+  blk_path = udisks_block_dup_device (ud_blk);
+
+out:
+  g_clear_object (&ud_blk);
+  g_clear_object (&ud_lx_blk_obj);
+  return blk_path;
+}
+
+static gboolean
+led_control (UDisksDriveLsmLocal *ud_drv_lsm_local,
+          GDBusMethodInvocation *invocation,
+          int (*lsm_func) (const char *blk_path, lsm_error **lsm_err),
+          const gchar *lsm_fun_name)
+{
+  lsm_error *lsm_err = NULL;
+  int lsm_rc = 0;
+  const gchar *blk_path = NULL;
+
+  blk_path = get_blk_path(ud_drv_lsm_local, invocation);
+  if (blk_path == NULL)
+    {
+      g_dbus_method_invocation_return_error
+        (invocation, UDISKS_ERROR, UDISKS_ERROR_FAILED,
+         "Failed to retrieve block path of specified disk drive");
+      goto out;
+    }
+
+  lsm_rc = lsm_func ((const char *) blk_path, &lsm_err);
+  if (lsm_rc != LSM_ERR_OK)
+    {
+      if (lsm_rc == LSM_ERR_NO_SUPPORT)
+        g_dbus_method_invocation_return_error
+          (invocation, UDISKS_ERROR, UDISKS_ERROR_NOT_SUPPORTED,
+           "Specified disk does not support this action");
+      else
+        g_dbus_method_invocation_return_error
+          (invocation, UDISKS_ERROR, UDISKS_ERROR_FAILED,
+           "%s failed %d: %s", lsm_fun_name, lsm_error_number_get (lsm_err),
+           lsm_error_message_get (lsm_err));
+      goto out;
+    }
+
+ out:
+  free ((gchar *) blk_path);
+  if (lsm_err != NULL)
+    lsm_error_free (lsm_err);
+
+  return TRUE; /* returning TRUE means that we handled the method invocation */
+}
+
+static gboolean
+handle_turn_ident_ledon (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                         GDBusMethodInvocation *invocation)
+{
+  return led_control(ud_drv_lsm_local, invocation,
+                     lsm_local_disk_ident_led_on, "lsm_local_ident_led_on");
+}
+
+static gboolean
+handle_turn_ident_ledoff (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                          GDBusMethodInvocation *invocation)
+{
+  return led_control(ud_drv_lsm_local, invocation,
+                     lsm_local_disk_ident_led_off, "lsm_local_ident_led_off");
+}
+
+static gboolean
+handle_turn_fault_ledon (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                         GDBusMethodInvocation *invocation)
+{
+  return led_control(ud_drv_lsm_local, invocation,
+                     lsm_local_disk_fault_led_on, "lsm_local_fault_led_on");
+}
+
+static gboolean
+handle_turn_fault_ledoff (UDisksDriveLsmLocal *ud_drv_lsm_local,
+                          GDBusMethodInvocation *invocation)
+{
+  return led_control(ud_drv_lsm_local, invocation,
+                     lsm_local_disk_fault_led_off, "lsm_local_fault_led_off");
+}
+
+static void
+udisks_linux_drive_lsm_local_finalize (GObject *object)
+{
+  udisks_debug ("LSM: udisks_linux_drive_lsm_local_finalize ()");
+
+  if (G_OBJECT_CLASS
+      (udisks_linux_drive_lsm_local_parent_class)->finalize != NULL)
+    G_OBJECT_CLASS
+      (udisks_linux_drive_lsm_local_parent_class)->finalize (object);
```

Comment:
I think this would be much clearer if it was only on two lines. We don't have to be so strict about line lengths these days, AFAICT.